storage/remote/queue_manager.go

```diff
@@ -430,6 +430,14 @@ func (s *shards) runShard(i int) {
 	pendingSamples := model.Samples{}
 
 	timer := time.NewTimer(s.qm.cfg.BatchSendDeadline)
+	defer func() {
+		if !timer.Stop() {
+			select {
+			case <-timer.C:
+			default:
+			}
```

Comment:
Nit, the Golang docs suggest a direct read from the timer, any reason to use this more verbose version?
if !t.Stop() {
	<-t.C
}
https://golang.org/pkg/time/#Timer.Stop

Reply:
This is to fix the bug that #3907 fixes - the `default:` option stops it stalling on an empty queue.

Reply:
It's a minor question, but this code is already rather complex. In what case is it unknown whether the timer channel is empty?